src/jit/valuenum.cpp

```diff
@@ -1988,13 +1954,23 @@ ValueNum ValueNumStore::VNForFunc(var_types typ, VNFunc func, ValueNum arg0VN, V
         {
             canFold = false;
         }
+
         if (typ == TYP_BYREF)
         {
             // We don't want to fold expressions that produce TYP_BYREF
             canFold = false;
         }
 
+        bool shouldFold = canFold;
+
         if (canFold)
+        {
+            // We can fold the expression, but we don't want to fold
+            // when the expression will always throw an exception
+            shouldFold = VNEvalShouldFold(typ, func, arg0VN, arg1VN);
```

Comment:
I have analyzed similar problem while writing folding check in importer and it seems that there are 2 possible scenarios with 2 possible decisions:
1. Expression tree is much bigger in comparison to tree with exception -> decision: it is better to fold.
2. Expression tree is comparable or smaller than exception tree -> decision: do not fold.

Reply:
Actually the reason that VNEvalShouldFold returns false is that the folding of some expressions would turn them into an unconditional throw.   When the expression unconditionally throws it does not produce a useful "normal" value number and using  a dummy value such a zero causes problems because we think that we can use that zero in the CSE phase. 

It is simpler to detect such cases and not try to fold them in value numbering.
So we do not fold (5 / 0) or (MIN_INT / -1)

Reply:
Got it. Thanks for explanation